### PR TITLE
fix: staking tx

### PR DESCRIPTION
**Pull request by jdkanani**

### CODE REVIEW

File: helper/config.go [hunk 1]
```diff
@@ -133,74 +129,59 @@ var GenesisDoc tmTypes.GenesisDoc
...
 
-	// load pv file, unmarshall and set to privObject
-	err = file.PermCheck(file.Rootify("priv_validator_key.json", configDir), secretFilePerm)
-	if err != nil {
-		Logger.Error(err.Error())
-	}
-	privVal := privval.LoadFilePV(filepath.Join(configDir, "priv_validator_key.json"), filepath.Join(configDir, "priv_validator_key.json"))
-	cdc.MustUnmarshalBinaryBare(privVal.Key.PrivKey.Bytes(), &privObject)
-	cdc.MustUnmarshalBinaryBare(privObject.PubKey().Bytes(), &pubObject)
+	// // load pv file, unmarshall and set to privObject
```
Comment (murthyvitwit) [inline helper/config.go:176]: Can we remove it if we are not using these lines(176-183)?

File: proto/staking/v1beta1/msg.proto [hunk 2]
```diff
@@ -84,12 +74,11 @@ message MsgSignerUpdate {
             "github.com/maticnetwork/heimdall/types.ValidatorID",
         (gogoproto.customname) = "ID"
     ];
-    google.protobuf.Any new_signer_pub_key = 3
-        [(gogoproto.casttype) =
-             "github.com/maticnetwork/heimdall/types/common.PubKey"];
-    bytes tx_hash = 4
-        [(gogoproto.casttype) =
-             "github.com/maticnetwork/heimdall/types/common.HeimdallHash"];
+    google.protobuf.Any new_signer_pub_key = 3 [
```
Comment (murthyvitwit) [inline proto/staking/v1beta1/msg.proto:77]: We thought of using string for pub_keys, isn't it? above (line 38) we just used singer_pub_key as a string.

File: types/common/hash.go [hunk 3]
```diff
@@ -74,13 +75,18 @@ func (aa *HeimdallHash) UnmarshalJSON(data []byte) error {
...
+func (aa *HeimdallHash) UnmarshalYAML(unmarshal func(interface {}) error) error {
 	var s string
-	err := yaml.Unmarshal(data, &s)
+	err := unmarshal(&s)
 	if err != nil {
 		return err
 	}
 
+	if s == "" {
+		aa = nil
```
Comment (murthyvitwit) [inline types/common/hash.go:86]: We can remove it as we returning immediately

File: x/chainmanager/types/genesis.go [hunk 4]
```diff
@@ -27,10 +27,10 @@ func NewGenesisState(params *Params) *GenesisState {
 	}
 }
 
-// // DefaultGenesisState - Return a default genesis state
-// func DefaultGenesisState() GenesisState {
-// 	return NewGenesisState(DefaultParams())
-// }
+// DefaultGenesisState - Return a default genesis state
```
Comment (murthyvitwit) [inline x/chainmanager/types/genesis.go:30]: We already have the **DefaultGenesis** method which is the same

File: x/staking/client/cli/tx.go [hunk 5]
```diff
@@ -241,7 +262,7 @@ func SignerUpdateTxCmd() *cobra.Command {
 			nonce, _ := cmd.Flags().GetUint64(FlagNonce)
 
 			// draft new SingerUpdate message
-			msg := types.NewMsgSignerUpdate(
+			msg, err := types.NewMsgSignerUpdate(
```
Comment (murthyvitwit) [inline x/staking/client/cli/tx.go:265]: Let's add the err validation

**Review by murthyvitwit — COMMENTED** [review]
I have added my comment, please check it out.

**Review by murthyvitwit — APPROVED** [review]

File: helper/config.go [hunk 6]
```diff
@@ -133,74 +129,51 @@ var GenesisDoc tmTypes.GenesisDoc
...
+		return err
 	}
 
 	maticClient = ethclient.NewClient(maticRPCClient)
+
 	// Loading genesis doc
 	genDoc, err := tmTypes.GenesisDocFromFile(filepath.Join(configDir, "genesis.json"))
 	if err != nil {
-		log.Fatal(err)
+		return nil
```
Comment (anilCSE) [inline helper/config.go:172]: Is it a typo? We shouldn't ignore this error here

File: proto/checkpoint/v1beta/query.proto [hunk 7]
```diff
@@ -16,101 +16,108 @@ option (gogoproto.unmarshaler_all) = true;
...
-  option (gogoproto.goproto_getters) = false;
+    option (gogoproto.equal)           = false;
+    option (gogoproto.goproto_getters) = false;
 
-  string bor_chainID = 1;
+    string bor_chainID = 1;
 }
 message QueryNextCheckpointResponse {
-  heimdall.checkpoint.v1beta1.MsgCheckpoint next_checkpoint = 1;
+    heimdall.checkpoint.v1beta1.MsgCheckpoint next_checkpoint = 1;
```
Comment (anilCSE) [inline proto/checkpoint/v1beta/query.proto:114]: We should add gogoproto annotations for multi-word keys. ref https://github.com/cosmos/cosmos-sdk/blob/v0.40.1/proto/cosmos/auth/v1beta1/auth.proto#L45

File: proto/heimdall/types/divident.proto [hunk 8]
```diff
@@ -14,8 +14,7 @@ message DividendAccount {
     option (gogoproto.goproto_getters)  = false;
     option (gogoproto.goproto_stringer) = false;
 
-    string user = 1
-        [(gogoproto.casttype) =
-        "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    string user       = 1 [(gogoproto.casttype) =
```
Comment (anilCSE) [inline proto/heimdall/types/divident.proto:17]: remove casttype to align with other definitions

File: proto/topup/v1beta/msg.proto [hunk 9]
```diff
@@ -22,9 +22,9 @@ service Msg {
 // GenesisState defines the topup module's genesis state.
 message MsgTopup {
     option (gogoproto.goproto_getters) = false;
-    string from_address = 1;
-    string user = 2;
-    uint64 fee = 3
+    string from_address                = 1;
```
Comment (anilCSE) [inline proto/topup/v1beta/msg.proto:25]: We should add gogoproto annotations for multi-word keys. ref https://github.com/cosmos/cosmos-sdk/blob/v0.40.1/proto/cosmos/auth/v1beta1/auth.proto#L45

File: x/staking/client/cli/tx.go [hunk 10]
```diff
@@ -45,6 +43,32 @@ func GetTxCmd() *cobra.Command {
...
 }
 
+func validateAndCompressPubKey(pubkeyBytes []byte) ([]byte, error) {
+	// convert PubKey to bytes
+	if len(pubkeyBytes) == 64 {
+		pubkeyBytes = helper.AppendPubkeyPrefix(pubkeyBytes)
+	}
+
+	// check if key is uncompressed
+	if len(pubkeyBytes) == 65 {
```
Comment (anilCSE) [inline x/staking/client/cli/tx.go:53]: Use constant

File: x/staking/client/cli/tx.go [hunk 11]
```diff
@@ -45,6 +43,32 @@ func GetTxCmd() *cobra.Command {
 	return stakingTxCmd
 }
 
+func validateAndCompressPubKey(pubkeyBytes []byte) ([]byte, error) {
+	// convert PubKey to bytes
+	if len(pubkeyBytes) == 64 {
```
Comment (anilCSE) [inline x/staking/client/cli/tx.go:48]: Use constant

File: x/staking/client/cli/tx.go [hunk 12]
```diff
@@ -45,6 +43,32 @@ func GetTxCmd() *cobra.Command {
 	return stakingTxCmd
 }
 
+func validateAndCompressPubKey(pubkeyBytes []byte) ([]byte, error) {
+	// convert PubKey to bytes
```
Comment (anilCSE) [inline x/staking/client/cli/tx.go:47]: mis-leading godoc

File: helper/util.go [hunk 13]
```diff
@@ -79,6 +80,37 @@ func EventByID(abiObject *abi.ABI, sigdata []byte) *abi.Event {
 	return nil
 }
 
+// AppendPubkeyPrefix returns publickey in uncompressed format
+func AppendPubkeyPrefix(signerPubKey []byte) []byte {
+	// append prefix - "0x04" as heimdall uses publickey in uncompressed format. Refer below link
+	// https://superuser.com/questions/1465455/what-is-the-size-of-public-key-for-ecdsa-spec256r1
+	prefix := make([]byte, 1)
```
Comment (anilCSE) [inline helper/util.go:87]: Should we ensure if the given input doesn't have the prefix or some other validation?

**Review by anilCSE — COMMENTED** [review]
I couldn't review this in-time but looks like there are some fixes required. How about a follow-up PR @jdkanani @murthyvitwit

File: helper/config.go (hunk 6 above)
Comment (jdkanani) [reply helper/config.go:172]: We are handling at cmd level - https://github.com/maticnetwork/heimdall/blob/a7384c5eb7affe10f4992c9963521eedd1cbb42e/cmd/heimdalld/cmd/root.go#L83-L85

File: proto/checkpoint/v1beta/query.proto (hunk 7 above)
Comment (jdkanani) [reply proto/checkpoint/v1beta/query.proto:114]: I don't understand. Can you help here? cc @murthyvitwit

File: helper/util.go (hunk 13 above)
Comment (jdkanani) [reply helper/util.go:87]: That validation is out of this method. This method just adds a prefix when gets called.

File: proto/checkpoint/v1beta/query.proto (hunk 7 above)
Comment (anilCSE) [reply proto/checkpoint/v1beta/query.proto:114]: We just need to change it to something like:
```go
heimdall.checkpoint.v1beta1.MsgCheckpoint next_checkpoint = 1 [(gogoproto.moretags) = "yaml:\"next_checkpoint\""];
```

Comment (murthyvitwit) [reply proto/checkpoint/v1beta/query.proto:114]: Sure, I will add them.

File: proto/heimdall/types/divident.proto (hunk 8 above)
Comment (murthyvitwit) [reply proto/heimdall/types/divident.proto:17]: This has been addressed in gov_porting PR.

File: proto/topup/v1beta/msg.proto (hunk 9 above)
Comment (murthyvitwit) [reply proto/topup/v1beta/msg.proto:25]: Copy

Comment (murthyvitwit) [reply proto/topup/v1beta/msg.proto:25]: Addressed in another PR

File: proto/heimdall/types/divident.proto (hunk 8 above)
Comment (murthyvitwit) [reply proto/heimdall/types/divident.proto:17]: Addressed in another PR (Gov)

File: proto/checkpoint/v1beta/query.proto (hunk 7 above)
Comment (murthyvitwit) [reply proto/checkpoint/v1beta/query.proto:114]: Addressed in another PR (Gov)

**Pull request merged**